### ISSUE TITLE
DDP with streaming got duplicate data

### ISSUE THREAD

**Issue by gongel** [mosaicml/streaming#324]
Writer
```python
import numpy as np
from streaming import MDSWriter

# Directory in which to store the compressed output files
data_dir = 'dirname'

# A dictionary mapping input fields to their data types
columns = {
    'image': 'ndarray',
    'class': 'int'
}

# Shard compression, if any
compression = 'zstd'

# Save the samples as shards using MDSWriter
with MDSWriter(out=data_dir, columns=columns, compression=compression) as out:
    for i in range(8):
        sample = {
            'image': np.random.randint(i+1, i+2, (1, 1)),
            'class': np.random.randint(i+1, i+2),
        }
        print(sample)
        out.write(sample)

```

reader.py
```python
from torch.utils.data import DataLoader
from streaming import StreamingDataset
import torch.distributed as dist

dist.init_process_group(backend = 'nccl')

# Remote path where full dataset is persistently stored
# remote = 's3://path-to-dataset'
remote = 'dirname'

# Local working dir where dataset is cached during operation
local = 'cache'

# Create streaming dataset
dataset = StreamingDataset(local=local, remote=remote, shuffle=False)

# Let's see what is in sample #1337...
sample = dataset[2]
img = sample['image']
cls = sample['class']

# Create PyTorch DataLoader
dataloader = DataLoader(dataset, batch_size=1)

for idx, i in enumerate(dataloader):
    if dist.get_rank() == 3:
        # import pdb;pdb.set_trace()
        print(f"Rank: {dist.get_rank()}, data: {i}, idx: {idx}")

```

Launch reader: 
```shell
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
python -m torch.distributed.launch --nproc_per_node=4 reader.py
```

Then, I will get output:
```shell
Rank: 3, data: {'class': tensor([1]), 'image': tensor([[[1]]])}, idx: 0
Rank: 3, data: {'class': tensor([1]), 'image': tensor([[[1]]])}, idx: 1
Rank: 3, data: {'class': tensor([2]), 'image': tensor([[[2]]])}, idx: 2
Rank: 3, data: {'class': tensor([2]), 'image': tensor([[[2]]])}, idx: 3
Rank: 3, data: {'class': tensor([3]), 'image': tensor([[[3]]])}, idx: 4
Rank: 3, data: {'class': tensor([3]), 'image': tensor([[[3]]])}, idx: 5
Rank: 3, data: {'class': tensor([4]), 'image': tensor([[[4]]])}, idx: 6
Rank: 3, data: {'class': tensor([4]), 'image': tensor([[[4]]])}, idx: 7
Rank: 3, data: {'class': tensor([5]), 'image': tensor([[[5]]])}, idx: 8
Rank: 3, data: {'class': tensor([5]), 'image': tensor([[[5]]])}, idx: 9
Rank: 3, data: {'class': tensor([6]), 'image': tensor([[[6]]])}, idx: 10
Rank: 3, data: {'class': tensor([6]), 'image': tensor([[[6]]])}, idx: 11
Rank: 3, data: {'class': tensor([7]), 'image': tensor([[[7]]])}, idx: 12
Rank: 3, data: {'class': tensor([7]), 'image': tensor([[[7]]])}, idx: 13
Rank: 3, data: {'class': tensor([8]), 'image': tensor([[[8]]])}, idx: 14
```

**Comment by yangwenz** [comment]
Found a similar bug with 0.5.0 and 0.5.1

**Comment by karan6181** [comment]
Hi @yangwenz , I think this is not a bug but the way we have the default value of the streaming dataset parameter. Can you try providing the 'num_canonical_node' value to 1 in StreamingDataset class or increase the number of samples to let say 100 or 500 ?

**Comment by yangwenz** [comment]
Thanks for the quick reply! I will tune the shuffling parameters.

**Comment by gongel** [comment]
Thx. It is right.
```python
dataset = StreamingDataset(local=local, remote=remote, shuffle=False, num_canonical_nodes=1)
```
